core/maximal_repeats.py

```diff
@@ -20,113 +23,123 @@ class Interval:
     def update_child_list(self, interval: Interval) -> None:
         self.child_list.append(interval)
 
-    def __str__(self):
+    def __str__(self) -> str:
         return f"lb={self.rb}, rb={self.rb}, lcp_value={self.lcp_value}"
 
 
-def add(top_interval: Interval, last_interval: Interval):
-    top_interval.update_child_list(last_interval)
-
-
 class SpecialStack:
-    def __init__(self):
+    def __init__(self) -> None:
         self.__stack = []
 
-    def top(self):
+    def top(self) -> Interval:
         if len(self.__stack) == 0:
             raise Exception("Trying to access top on an empty stack")
         else:
             return self.__stack[-1]
 
-    def pop(self):
+    def pop(self) -> Interval:
         return self.__stack.pop()
 
-    def push(self, interval: Interval):
+    def push(self, interval: Interval) -> None:
         self.__stack.append(interval)
 
-    def is_empty(self):
+    def is_empty(self) -> bool:
         return len(self.__stack) == 0
 
 
-def process(lcp_interval: Interval, suf_tab, s):
-    pos_set_list_head = None
-    prev = None
-
-    pos_set = []
+def process(lcp_interval: Interval, suf_tab: list[int], s: str, bwt_table: list[str]) -> None:
+    # linked list to union the position sets and save it to the parent node(propagate)
+    pos_set_list_head: Optional[Node] = None
+    prev: Optional[Node] = None
+    # init the alphabet set over the input string
+    alphabet_set: set[str] = set(s)
+    # init the dict that maps the alphabet to a pos p, -1 is for undefined at init
+    alphabet_pos_map: dict[str, int] = {alphabet: 0 for alphabet in alphabet_set}
     lcp = lcp_interval.lcp_value
 
     for j in range(0, len(lcp_interval.child_list)):
```

Comment:
The construction `range(0, len(x))` is considered bad form. First, because `range(0, n)` is the same as `range(n)` and second because of reasons I don't know, but some people have a strong opinion about it. Probably because in most cases where you use `for i in range(len(x))` you don't use `i` anywhere except for in `x[i]` so `for a in x` would be faster and less likely to have bugs. You don't use `j` here, as far as I can tell, except for in `lcp_interal.child_list[j]` so you could have used `for child in lcp_interval.child_list` instead.